libflux/go/libflux/parser.go

```diff
@@ -0,0 +1,49 @@
+// +build libflux
+
+package libflux
+
+// #cgo CFLAGS: -I${SRCDIR}/../../include
+// #cgo LDFLAGS: -L. -lflux
+// #include <influxdata/flux.h>
+// #include <stdlib.h>
+import "C"
+
+import (
+	"runtime"
+	"unsafe"
+)
+
+// freeable indicates a resource that has memory
+// allocated to it outside of Go and must be freed.
+type freeable interface {
+	Free()
+}
+
+// free is a utility method for calling Free
+// on a resource.
+func free(f freeable) {
+	f.Free()
+}
+
+// ASTFile is a parsed AST.
+type ASTFile struct {
+	ptr *C.struct_flux_ast_t
```

Comment:
How did this `struct_flux_ast_t` map to `ast::File` in Rust?

Reply:
Should be the flat buffers serialization in the future I guess 🤔

Reply:
OK so in the PR it is just a test run and we do not look into any details in the struct.

Reply:
Yeah, I think that @jsternberg wants to show us that we can call Rust from Go and properly free memory.

He demonstrates that by only using a pointer to something. In the future that should be a fbs structure. I am only guessing though and I was curious to have a look at this PR :)

Reply:
It maps to this definition in the C file.

```
struct flux_ast_t;
```

The C binding generated by Go will map that to the `C.struct_flux_ast_t` type. Since the struct is a forward declaration, the only thing it can be used for is a pointer. When you're writing C bindings, this is common to create handles that can be passed back into the library and to avoid leaking the internal details of the implementation since the attributes in the struct become part of the public ABI.

Reply:
To clarify, the implementation is very likely to always be an empty struct. It may be possible in certain circumstances to expose a struct's inner details. It is possible to tell Rust how to arrange the memory so the C definition and the Rust definition share the same memory layout, but it is probably not something we should use too frequently.

We would use the handles to access the data. See this [gist](https://gist.github.com/jsternberg/3adac9a7b3386d7eb53846e47955604b#sharing-memory) for a rough outline of what I mean.